Users can edit and cancel a void booking

patrickjfl opened the pull request:
# Context

Allows users to [edit and cancel a void booking](https://trello.com/c/AYpNNB2q/757-users-can-edit-and-cancel-a-void-booking). For now the changes remain feature flagged and disabled outside of dev, test and local environments.

End to end testing is not included in this PR but will follow shortly.

<!-- Is there a Trello ticket you can link to? -->
<!-- Do you need to add any environment variables? -->
<!-- Is an ADR required? An ADR should be added if this PR introduces a change to the architecture. -->

# Changes in this PR

## Screenshots of UI changes

View void booking screen, with edit/cancel actions:
![Screenshot 2023-03-14 at 11 08 36](https://user-images.githubusercontent.com/70749355/224983529-0bbff940-e374-4a0a-9c3d-3995e041bad3.png)

Edit void booking screen:
![Screenshot 2023-03-14 at 11 09 27](https://user-images.githubusercontent.com/70749355/224983681-162e5bc8-2ffc-4dff-bec4-2ceb1098f4e0.png)

View void booking screen, with update success message:
![Screenshot 2023-03-14 at 11 09 41](https://user-images.githubusercontent.com/70749355/224983774-1e7580d2-9669-4f05-8636-089eb2666875.png)

Cancel void booking screen:
![Screenshot 2023-03-14 at 11 10 04](https://user-images.githubusercontent.com/70749355/224983898-625a57d7-316a-4f46-82db-48a5a791dc0f.png)

View void booking screen, with cancellation success message and cancelled status:
![Screenshot 2023-03-14 at 11 10 10](https://user-images.githubusercontent.com/70749355/224983977-1696eb84-41dd-4485-9ab5-251d8abcfed7.png)

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Have all changes to any dependencies been deployed to both `preprod` and
    `production` in a backwards compatible way? (This includes our API,
    infrastructure, third-party integrations and any secrets or environment variables)
- [ ] Has a required data migration been included in this change or been done in
    advance?

## Post merge checklist

Once we've merged we now need to release this through to production before
considering it done.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-temporary-accommodation-ui)
and work through the following steps:

- [ ] Has the product manager asked to provide approval for this feature?
  - [ ] Has the product manager approved?
- [ ] Manually approve release to preprod
- [ ] Verify change released to preprod
- [ ] Manually approve release to prod
- [ ] Verify change released to production

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

Commit:
Add cancel and update methods to lostBedClient and wiremock

We add update and cancel methods to lostBedClient, together with updated test factories and wiremock stubs.

Commit:
Adds cancel and update methods to lostBedService

We update lostBedService with updateLostBed and cancelLostBed methods.

We create paths and add methods to lostBedsController to support editing and cancelling lost beds. We also create nunjucks templates to render these screens.

Commit:
Route controller

We add routing to make controller methods reachable by URL.

Commit:
Add integration tests for edit/cancel lost beds

patrickjfl commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Users can edit and cancel a void booking](https://trello.com/c/AYpNNB2q/757-users-can-edit-and-cancel-a-void-booking)

bryangaledxw commented on server/@types/shared/models/LostBedCancellation.ts:
```diff
@@ -3,8 +3,7 @@
+  id: string
+  createdAt: string
+  notes?: string
+}
```
Able to rerun `generate-types`, and squash the result into this commit? Just to keep `server/@types` as the unmodified output from that script

bryangaledxw commented on server/services/lostBedService.ts:
```diff
@@ -29,6 +32,44 @@ export default class LostBedService {
+
+    return lostBedCancellation
+  }
+
```
`getUpdateLostBed` makes sense as the method is as much about the type conversion as anything else. `updateLostBed` and `cancelLostBed`, I'd be tempted to rename to just `update` and `cancel`

bryangaledxw commented on server/paths/api.ts:
```diff
@@ -5,7 +5,7 @@ const singlePremisesPath = premisesPath.path(':premisesId')
 const lostBedsPath = singlePremisesPath.path('lost-beds')
 const singleLostBedPath = lostBedsPath.path(':lostBedId')
-const cancelLostBedPath = singleLostBedPath.path('cancellation')
+const cancelLostBedPath = singleLostBedPath.path('cancellations')
```
Think this belongs in the first commit

patrickjfl replied:
ah thanks :)

bryangaledxw commented on server/paths/temporary-accommodation/manage.ts:
```diff
@@ -67,6 +67,9 @@ const paths = {
     show: singleLostBedPath,
+    edit: singleLostBedPath.path('edit'),
+    update: singleLostBedPath,
+    cancellations: { new: singleLostBedPath.path('cancel'), create: singleLostBedPath.path('cancellations') },
```
I feel like you've probably considered this and decided against, but my instinct here would be to keep this constent with bookings, so:

```
cancellations: { new: lostBedCancellationsPath.path('new'), create: lostBedCancellationsPath },
```

patrickjfl replied:
Yeah, I see your point, I think it just feels kind of misleading to/confusing for users if the URL implies that they can make multiple cancellations of a single void.

bryangaledxw commented on server/views/temporary-accommodation/lost-beds/cancel.njk:
```diff
@@ -0,0 +1,40 @@
+            text: "Submit"
+        }) }}
+  </form>
+{% endblock %}
```
Let's end with a newline - this always looks a like a warning to me. And I'd prefer some vertical whitespace between the sections. Same for `edit.njk`

patrickjfl replied:
ah yeah, nunjucks again. How do you format yours?

bryangaledxw replied:
I feel like at a minimum, after the `{% from ... %}`s, after the `{% extends ... %}`, and before any `{% block %}`s, and between the "header" bits of the page and anything else.

I did go down the rabbit hole of automatic Nunjucks linting on my last dxw project, but couldn't find anything that wasn't bad in some way or another!

bryangaledxw commented on server/views/temporary-accommodation/lost-beds/cancel.njk:
```diff
@@ -0,0 +1,40 @@
+{% set pageTitle = applicationName + " - Cancel void booking" %}
+{% set mainClasses = "app-container govuk-body" %}
+{% block content %}
+  {{ breadCrumb('Cancel a void booking', [
+    {title: 'List of bedspaces', href: paths.premises.show({ premisesId: premises.id })},
+    {title: 'View bedspace', href: paths.premises.bedspaces.show({ premisesId: premises.id, roomId: room.id })}
+  ]) }}
```
I don't know if I can justify my intuition here, but I feel like this is a `govukBackLink`, rather than `breadCrumb`. Here and for `edit.njk` My mental model is that there's a tree of pages, then things like this and booking state changes are these odd little ephemeral spindles off of that. Like, they're definitely not permanent parts of the tree, they're only here for particular states.

I think maybe that's my rule - "permanent pages", pages that might make sense to bookmark, breadcrumbs. Otherwise, it wants a backlink

bryangaledxw commented on server/controllers/temporary-accommodation/manage/lostBedsController.ts:
```diff
@@ -81,8 +85,121 @@ export default class LostBedsController {
+        )
+      }
+    }
+  }
```
Can go either way, since I don't think this is going to expand to a bunch of different actions like bookings, but consider if it's worth pulling these two methods into their own controller

bryangaledxw commented on server/controllers/temporary-accommodation/manage/lostBedsController.ts:
```diff
@@ -81,8 +85,121 @@ export default class LostBedsController {
     }
   }
+
+  update(): RequestHandler {
```
Would swap the order of `update` and `edit` here and in the test, just to reflect user journey order

bryangaledxw commented on server/components/lostBedInfo.ts:
```diff
@@ -26,7 +26,10 @@ export default (lostBed: LostBed): SummaryList['rows'] => {
   rows.push({
     key: textValue('Notes'),
-    value: htmlValue(formatLines(lostBed.notes)),
+    value:
+      lostBed.status === 'active'
+        ? htmlValue(formatLines(lostBed.notes))
+        : htmlValue(formatLines(lostBed.cancellation.notes)),
```
Suggest reflect this change in `lostBedInfo.test.ts`

bryangaledxw commented on server/routes/temporary-accommodation/manage.ts:
```diff
@@ -200,6 +200,35 @@ export default function routes(controllers: Controllers, services: Services, rou
+        },
+        {
+          path: paths.lostBeds.show.pattern,
+          auditEvent: 'CREATE_LOST_BED_SUCCESS',
```
These should be unique

patrickjfl replied:
oh god, good spot. I think that got mixed up in rebasing

bryangaledxw commented on integration_tests/tests/temporary-accommodation/manage/lostBed.cy.ts:
```diff
@@ -197,14 +201,34 @@ context('Lost bed', () => {
     const premises = premisesFactory.build()
     const room = roomFactory.build()
-    const lostBed = lostBedFactory.build()
+    const lostBed = lostBedFactory.build({ status: 'active' })
```
Definitely not a blocker, but consider expanding the factory so we can just do `lostBedFactory.active().build()` here

patrickjfl replied:
huh, I already did that but just forgot about it! Will amend

bryangaledxw commented on integration_tests/tests/temporary-accommodation/manage/lostBed.cy.ts:
```diff
@@ -244,4 +268,153 @@ context('Lost bed', () => {
+
+      expect(requestBody.notes).equal(cancelLostBed.notes)
+    })
+  })
```
May be worth adding a couple of tests to check that going back from cancel/edit go to the right place

patrickjfl replied:
good call

bryangaledxw commented on server/services/lostBedService.test.ts:
```diff
@@ -10,6 +10,9 @@ import ReferenceDataClient from '../data/referenceDataClient'
 import lostBedFactory from '../testutils/factories/lostBed'
```
Check the message in this commit - think that comma wants to be a full stop

bryangaledxw reviewed (approved):
Good stuff! Some feedback, but only must-fix is the audit event names.

I was trying a failing to decide if I'd put the new controller method in their own controllers. It feel like we should do the same thing with both edit and cancel. But then edit feels like it belongs in the existing controller, but cancel maybe feels more like it belongs on its own, at least to be consistent with bookings. Anyway, I'm good for it to stay as is if you're happy with it.

I think don't worry about it for this PR, but for something like this, I'd maybe have added update (and testing that) as one block of commits, then added cancel (and its testing), so we're adding a new thing and its testing as close together as we can

Commit:
Extend lostBedsController to support editing/cancelling

We create paths and add methods to lostBedsController to support editing and cancelling lost beds, We also create nunjucks templates to render these screens.

We also amend lostBedInfo component to conditionally render lost bed notes based on the status of the lost bed.
